File: engine/src/main/java/com/terralite/engine/time/GameClock.java
package com.terralite.engine.time;

import java.time.Duration;
import java.util.Objects;

public final class GameClock {
    private final Duration dayLength;
    private Duration elapsed = Duration.ZERO;

    public GameClock(Duration dayLength) {
        if (Objects.requireNonNull(dayLength, "dayLength").isZero() || dayLength.isNegative()) {
            throw new IllegalArgumentException("Day length must be positive");
        }
        this.dayLength = dayLength;
    }

    public static GameClock defaultClock() {
        return new GameClock(Duration.ofMinutes(20));
    }

    public Duration dayLength() {
        return dayLength;
    }

    public Duration elapsed() {
        return elapsed;
    }

    public long day() {
        return Math.floorDiv(elapsed.toNanos(), dayLength.toNanos());
    }

    public double dayProgress() {
        return (elapsed.toNanos() % dayLength.toNanos()) / (double) dayLength.toNanos();
    }

    public void advance(Duration delta) {
        Objects.requireNonNull(delta, "delta");
        if (delta.isNegative()) {
            throw new IllegalArgumentException("Clock delta cannot be negative");
        }
        elapsed = elapsed.plus(delta);
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/time/GameClock.kt
package com.terralite.engine.time

/**
 * Tracks in-game time. All durations are in nanoseconds.
 */
class GameClock(private val dayLengthNanos: Long) {
    private var elapsedNanos: Long = 0L

    init {
        require(dayLengthNanos > 0) { "Day length must be positive" }
    }

    /** @return day length in nanoseconds */
    fun dayLength(): Long = dayLengthNanos

    /** @return elapsed time in nanoseconds */
    fun elapsed(): Long = elapsedNanos

    fun day(): Long = Math.floorDiv(elapsedNanos, dayLengthNanos)

    fun dayProgress(): Double = (elapsedNanos % dayLengthNanos).toDouble() / dayLengthNanos.toDouble()

    fun advance(deltaNanos: Long) {
        require(deltaNanos >= 0) { "Clock delta cannot be negative" }
        elapsedNanos += deltaNanos
    }

    companion object {
        private const val TWENTY_MINUTES_NANOS = 20L * 60L * 1_000_000_000L

        @JvmStatic fun defaultClock(): GameClock = GameClock(TWENTY_MINUTES_NANOS)
    }
}
